pmm-app/src/pmm-settings/Settings.tsx

```diff
@@ -0,0 +1,119 @@
+import React, {
+  FC,
+  useEffect,
+  useMemo,
+  useState
+} from 'react';
+import {
+  Spinner,
+  TabContent,
+  Tab,
+  useTheme
+} from '@grafana/ui';
+import { cx } from 'emotion';
+import { TabsVertical } from 'shared/components/Elements/TabsVertical/TabsVertical';
+import { Diagnostics } from './components/Diagnostics/Diagnostics';
+import { MetricsResolution } from './components/MetricsResolution/MetricsResolution';
+import { SSHKey } from './components/SSHKey/SSHKey';
+import { AlertManager } from './components/AlertManager/AlertManager';
+import { Advanced } from './components/Advanced/Advanced';
+import { SettingsService, LoadingCallback } from './Settings.service';
+import { Settings, TabKeys } from './Settings.types';
+import { Messages } from './Settings.messages';
+import { getSettingsStyles } from './Settings.styles';
+
+
+export const SettingsPanel: FC = () => {
+  const theme = useTheme();
+  const styles = getSettingsStyles(theme);
+  const [activeTab, setActiveTab] = useState('metrics');
+  const {
+    metrics,
+    advanced,
+    ssh,
+    alertManager
+  } = Messages.tabs;
+  const tabs = useMemo(
+    () => [
+      { label: metrics, key: TabKeys.metrics, active: activeTab === TabKeys.metrics },
+      { label: advanced, key: TabKeys.advanced, active: activeTab === TabKeys.advanced },
+      { label: ssh, key: TabKeys.ssh, active: activeTab === TabKeys.ssh },
+      { label: alertManager, key: TabKeys.alertManager, active: activeTab === TabKeys.alertManager },
+    ],
+    [activeTab]
+  );
+  const [settings, setSettings] = useState<Settings>();
+  const [loading, setLoading] = useState(true);
+  const updateSettings = async (body: any, callback: LoadingCallback) => {
+    const response = await SettingsService.setSettings(body, callback);
+
+    if (response) {
+      setSettings(response);
+    }
+  };
+
+  useEffect(() => {
+    SettingsService.getSettings(setLoading, setSettings);
```

Comment:
I don't really like this approach, is there a particular reason for using callbacks here? I find this a bit obscure.

Reply:
Yes, the callbacks were not to block the main execution waiting for the request and don't have to bring those try/catch blocks here. 
I was using the `SettingsService.setSettings` on each separated component at first, and having to replicate the same error handling code and loading was too much duplication. And the API layer is not dependent on the model because it defines the the callback interfaces that it expects (so it depends on an abstraction and not concrete objects).
I saw the other comments and yes I agree that the presentation layer shouldn't know nothing about the requests, maybe by having a middle layer to transform the model to API or vice versa. Or use any external library, but we definitely have to work on that.